extensions/amp-web-push/0.1/web-push-service.js

```diff
@@ -0,0 +1,697 @@
+        null
+    );
+  }
+
+  queryServiceWorkerState_() {
+    return this.queryHelperFrame_(
+        WindowMessenger.Topics.SERVICE_WORKER_STATE,
+        null
+    );
+  }
+
+  registerServiceWorker() {
+    return this.queryHelperFrame_(
+        WindowMessenger.Topics.SERVICE_WORKER_REGISTRATION,
+        {
+          workerUrl: this.config_.serviceWorkerUrl,
+          registrationOptions: this.config_.serviceWorkerRegistrationOptions ||
+          {scope: '/'},
+        }
+    );
+  }
+
+  querySubscriptionStateRemotely_() {
+    return this.queryServiceWorker_(
+        'amp-web-push-subscription-state',
+        null
+    );
+  }
+
+  subscribeForPushRemotely_() {
+    return this.queryServiceWorker_(
+        'amp-web-push-subscribe',
+        null
+    );
+  }
+
+  unsubscribeFromPushRemotely_() {
+    return this.queryServiceWorker_(
+        'amp-web-push-unsubscribe',
+        null
+    );
+  }
+
+  isServiceWorkerActivated_() {
+    const self = this;
+    return this.queryServiceWorkerState_().then(
+      /** @param {{isControllingFrame:boolean, state:string, url:string}}
+          serviceWorkerState */
+        function(serviceWorkerState) {
+          const isControllingFrame =
+            serviceWorkerState.isControllingFrame === true;
+          const serviceWorkerHasCorrectUrl =
+            serviceWorkerState.url === self.config_.serviceWorkerUrl;
+          const serviceWorkerActivated =
+          serviceWorkerState.state === 'activated';
+
+          return isControllingFrame &&
+          serviceWorkerHasCorrectUrl &&
+          serviceWorkerActivated;
+        });
+  }
+
+  /*
+    Sets the visibilities of subscription and unsubscription
+    <amp-web-push> elements.
+
+    Element visibilities change throughout the lifetime of the page: they are
+    initially invisible as their visibilties are determined, and then they
+    either remain hidden or appear. After users subscribe or unsubscribe,
+    visibilties change again.
+  */
+  setWidgetVisibilities(widgetCategoryName, isVisible) {
+    const widgetDomElements = this.ampdoc.getRootNode()
+      .querySelectorAll(`${TAG}[visibility=${widgetCategoryName}]`);
+    const visibilityCssClassName = 'invisible';
+
+    for (let i = 0; i < widgetDomElements.length; i++) {
+      const widgetDomElement = widgetDomElements[i];
+      if (isVisible) {
+        widgetDomElement.classList.remove(visibilityCssClassName);
+      } else {
+        widgetDomElement.classList.add(visibilityCssClassName);
+      }
+    }
+  }
+
+  getSubscriptionStateReplyVersion_(subscriptionStateReply) {
+    if (typeof subscriptionStateReply === 'boolean') {
+      return 1;
+    }
+  }
+
+  updateWidgetVisibilities() {
+    return this.queryNotificationPermission_().then(notificationPermission => {
+      if (notificationPermission === NotificationPermission.DENIED) {
+        this.updateWidgetVisibilitiesNotificationPermissionsBlocked();
```

Comment:
SG. Generally, though, I'm very much ok with using private methods for testing, especially if they are documented as such.